### PR TITLE
fix(client): worker executor task once

### LINKED ISSUE #46949: `task.once` is invalid in worker-executor.js
## Describe the bug

`task.once` is invalid in `client/src/templates/Challenges/utils/worker-executor.js`.

## To Reproduce

1. add a test in `client/src/templates/Challenges/utils/worker-executor.test.js`

   ```js
   it('Task should emit handler once', () => {
     mockWorker();
     const testWorker = createWorker('test');
     const task = testWorker.execute('test');
     const handler = jest.fn();
     task.once('testOnce', handler);
   
     task.emit('testOnce', handler);
     task.emit('testOnce', handler);
     expect(handler).toBeCalledTimes(1);
   });
   ```

2. run this test

3. If `task.once` works well, then `handler` will be called just once.

4. However, `handler` is called twice time.

## Expected behavior

`handler` should only be called once.

## Analyze

There is only one parameter `handler` in the function `self.removeListener` on Line 141

https://github.com/freeCodeCamp/freeCodeCamp/blob/8f19b1926bc51c0149254fdd7a86ecb0708c6dd2/client/src/templates/Challenges/utils/worker-executor.js#L141

```js
self.once = (event, listener) => {
  self.on(event, function handler(...args) {
    self.removeListener(handler);
    listener.apply(self, args);
  });
  return self;
};
```

But function`removeListener` needs two parameters (`event` and `listener`), since `handler` is type of function instead of type of string, so the program never enters the if judgment.

```js
  self.removeListener = (event, listener) => {
    if (typeof self._events[event] !== 'undefined') {
      const index = self._events[event].indexOf(listener);
      if (index !== -1) {
        self._events[event].splice(index, 1);
      }
    }
    return self;
  };
```

## Solution

Perhaps Line 141

https://github.com/freeCodeCamp/freeCodeCamp/blob/8f19b1926bc51c0149254fdd7a86ecb0708c6dd2/client/src/templates/Challenges/utils/worker-executor.js#L141

should be changed to

```js
self.removeListener(event, handler);
```

and change Line 96

https://github.com/freeCodeCamp/freeCodeCamp/blob/8f19b1926bc51c0149254fdd7a86ecb0708c6dd2/client/src/templates/Challenges/utils/worker-executor.js#L96 to

```js
task.done = new Promise((resolve, reject) => {
  task
    .on('done', data => resolve(data))
    .on('error', err => reject(err.message));
});
```

### PR DESCRIPTION
**Pull request by masterX89**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #46949

### CODE REVIEW

File: client/src/templates/Challenges/utils/worker-executor.js [hunk 1]
```diff
@@ -129,16 +129,23 @@ const eventify = self => {
 
   self.emit = (event, ...args) => {
     if (typeof self._events[event] !== 'undefined') {
-      self._events[event].forEach(listener => {
+      let length = self._events[event].length;
+      for (let i = 0; i < length; i++) {
+        const listener = self._events[event][i];
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/worker-executor.js:134]:
```suggestion
      const listeners = self._events[event].slice();
      for (let listener of listeners) {
```
The real issue is the splice, but we can work around it by copying the listener array.

File: client/src/templates/Challenges/utils/worker-executor.js [hunk 2]
```diff
@@ -129,16 +129,23 @@ const eventify = self => {
...
         listener.apply(self, args);
-      });
+        // may remove Listener so should change index and length
+        if (self._events[event].length < length) {
+          i--;
+          length--;
+        }
+      }
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/worker-executor.js:141]:
```suggestion
      }
```
Completes above suggestion.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
@masterX89 thanks for digging into this!  I think we can tweak your solution to make it a little clearer what's really happening:

**Comment by masterX89** [comment]
Thanks for suggestions, I changed it to what the comment suggested, it makes the function much more clear😊

**Comment by ojeytonwilliams** [comment]
@masterX89 it was only due to your efforts that I understood what was going on. Thanks again.

I can't see how your fix to `.once` could break anything, but I'll give this a proper QA tomorrow.

**Review by ojeytonwilliams — APPROVED** [review]
I tried it out on a few JS challenges and all seems to be fine. Nice work, @masterX89!

**Comment by masterX89** [comment]
Hi, @moT01 @Sboonny,
I think this issue, if fixed, could help resolve the TypeScript Migration for /client/src/templates/Challenges/utils/worker-executor.js and worker-executor.test.js
When you have time could you please take a look at this PR?
Thank you in advance :)

**Comment by Sboonny** [comment]
@masterX89 I haven't interacted with that file before, but I will pass it to our chat to give it some spot light which it deserves

**Comment by masterX89** [comment]
Thank you very much~😊

**Review by moT01 — APPROVED** [review]
LGTM @masterX89 🎉 Thanks for contributing to freeCodeCamp 🎉

**Pull request merged**